src/motion_lib/src/bicycle_model.cpp

```diff
@@ -0,0 +1,134 @@
+  double rl_velocity = common_lib::maths::get_wheel_velocity_from_rpm(rl_rpm, WHEEL_DIAMETER);
+  double rr_velocity = common_lib::maths::get_wheel_velocity_from_rpm(rr_rpm, WHEEL_DIAMETER);
+  if (steering_angle == 0) {  // If no steering angle, moving straight
+    velocities.first = (rl_velocity + rr_velocity) / 2;
+  } else if (steering_angle > 0) {
+    double rear_axis_center_rotation_radius = WHEELBASE / tan(steering_angle);
+    velocities.second = rl_velocity / (rear_axis_center_rotation_radius - (AXIS_LENGTH / 2));
+    velocities.first =
+        sqrt(pow(rear_axis_center_rotation_radius, 2) + pow(REAR_AXIS_TO_CAMERA, 2)) *
+        fabs(velocities.second);
+  } else {
+    double rear_axis_center_rotation_radius = WHEELBASE / tan(steering_angle);
+    velocities.second = rr_velocity / (rear_axis_center_rotation_radius + (AXIS_LENGTH / 2));
+    velocities.first =
+        sqrt(pow(rear_axis_center_rotation_radius, 2) + pow(REAR_AXIS_TO_CAMERA, 2)) *
+        fabs(velocities.second);
+  }
+  return velocities;
+}
+
+common_lib::structures::Position cg_2_rear_axis(common_lib::structures::Position cg,
+                                                double orientation, double dist_cg_2_rear_axis) {
+  common_lib::structures::Position rear_axis;
+  rear_axis.x = cg.x - dist_cg_2_rear_axis * cos(orientation);
+  rear_axis.y = cg.y - dist_cg_2_rear_axis * sin(orientation);
+  return rear_axis;
+}
+
+Eigen::VectorXd estimate_observations(Eigen::Vector3d& state, double wheel_base,
+                                      double weight_distribution_front, double gear_ratio,
+                                      double wheel_radius) {
+  double Lr = wheel_base *
+              weight_distribution_front;  // distance from the center of mass to the rear wheels
+  double Lf =
+      wheel_base *
+      (1 - weight_distribution_front);  // distance from the center of mass to the front wheels
+
+  double rear_wheel_velocity = sqrt(pow(state(0), 2) + pow(state(1) - state(2) * Lr, 2));
+  double rear_wheels_rpm = 60 * rear_wheel_velocity / (2 * M_PI * wheel_radius);
+  if (state(0) < 0) {
+    rear_wheels_rpm = -rear_wheels_rpm;
+  }
+
+  double front_wheel_velocity = sqrt(pow(state(0), 2) + pow(state(1) + state(2) * Lf, 2));
+  double front_wheels_rpm = 60 * front_wheel_velocity / (2 * M_PI * wheel_radius);
+  if (state(0) < 0) {
+    front_wheels_rpm = -front_wheels_rpm;
+  }
+
+  double steering_angle =
+      (std::fabs(state(0)) <= 0.01) ? 0 : atan((state(1) + state(2) * Lf) / state(0));
+  double motor_rpm = 60 * gear_ratio * rear_wheel_velocity / (2 * M_PI * wheel_radius);
+  if (state(0) < 0) {
+    motor_rpm = -motor_rpm;
+  }
+
+  Eigen::VectorXd observations = Eigen::VectorXd::Zero(6);
+  observations << front_wheels_rpm, front_wheels_rpm, rear_wheels_rpm, rear_wheels_rpm,
+      steering_angle, motor_rpm;
+  return observations;
+}
+
+Eigen::MatrixXd jacobian_of_observation_estimation(Eigen::Vector3d& state, double wheel_base,
```

Comment:
I am really confused, where did you get these formulas from? Maybe we should review this in person